client/comms/wsconn.go

```diff
@@ -0,0 +1,349 @@
+			log.Errorf("unable to set read deadline: %v", err)
+			return err
+		}
+
+		// Respond with a pong.
+		return conn.ws.WriteControl(websocket.PongMessage, []byte{},
+			now.Add(conn.cfg.WriteWait))
+	})
+
+	return nil
+}
+
+// keepAlive maintains an active websocket connection by reconnecting when
+// the established connection is broken. This should be run as a goroutine.
+func (conn *WsConn) keepAlive(ctx context.Context) {
+	ticker := time.NewTicker(time.Second)
+
+	for {
+		select {
+		case <-ticker.C:
+			if !conn.isConnected() {
+				err := conn.connect()
+				if err != nil {
+					log.Errorf("unable to establish a connection: %v", err)
+				}
+
+				// Run the reconnect sync after reestablishing a connection.
+				if conn.cfg.ReconnectSync != nil {
+					conn.cfg.ReconnectSync()
+				}
+			}
+
+		case <-ctx.Done():
+			now := time.Now()
+			closeMsg := websocket.FormatCloseMessage(websocket.CloseMessage, "")
+			conn.writeMtx.Lock()
+			err := conn.ws.WriteControl(websocket.CloseMessage, closeMsg,
+				now.Add(conn.cfg.WriteWait))
+			conn.writeMtx.Unlock()
+			if err != nil {
+				log.Errorf("unable to close websocket connection: %v", err)
+			}
+			ticker.Stop()
+			conn.wg.Done()
+			return
+		}
+	}
+}
+
+// read fetches and parses incoming messages for processing. This should be
+// run as a goroutine.
+func (conn *WsConn) read() {
+	for {
+		// Only allow reads if there is an established connection.
+		if !conn.isConnected() {
+			time.Sleep(time.Millisecond * 500)
+			continue
+		}
+
+		msg := new(msgjson.Message)
+		err := conn.ws.ReadJSON(msg)
+		if err != nil {
+			if opErr, ok := err.(*net.OpError); ok {
+				if opErr.Op == "read" {
+					conn.setConnected(false)
+					continue
+				}
+			}
+
+			if websocket.IsCloseError(err, websocket.CloseAbnormalClosure,
+				websocket.CloseGoingAway, websocket.CloseNormalClosure) {
+				return
+			}
+
+			if nErr, ok := err.(net.Error); ok {
+				if nErr.Timeout() {
+					conn.setConnected(false)
+					continue
+				}
+			}
+
+			if _, ok := err.(*json.UnmarshalTypeError); ok {
+				log.Error(err)
+				continue
+			}
+
+			log.Errorf("unable to read message: %v", err)
+			conn.cfg.Cancel()
+			return
+		}
+
+		conn.readCh <- msg
+	}
+}
+
+// send pushes outgoing messages over the websocket connection.
+// This should be run as a goroutine.
+func (conn *WsConn) send(ctx context.Context) {
+	for {
+		select {
+		case msg := <-conn.sendCh:
+			// Log the route if the message being sent is a request.
+			if msg.Type == msgjson.Request {
+				conn.logRequest(msg.ID, msg)
+			}
+
+			conn.writeMtx.Lock()
+			_ = conn.ws.SetWriteDeadline(time.Now().Add(conn.cfg.WriteWait))
+			err := conn.ws.WriteJSON(msg)
+			conn.writeMtx.Unlock()
+			if err != nil {
+				log.Errorf("unable to write message: %v", err)
+				continue
+			}
+
+		case <-ctx.Done():
+			conn.wg.Done()
+			return
+		}
+	}
+}
+
+// Run starts all processes of the websocket connection.
+func (conn *WsConn) Run(ctx context.Context) {
```

Comment:
Just a thought: would it be inappropriate for the constructor to call `Run`, which would let `ctx` be a field of `WsConn`?

Reply:
Yeah that would go against how contexts are supposed to be used. https://golang.org/pkg/context/ states contexts are to be passed where needed instead of making it a field of a struct.

Reply:
Context is a versatile type.  A common use is in fact a request-scoped object, where the functions getting the context did not make the context.  It is idiomatic to have it passed as an argument in most cases, but not if it makes no sense to pass it to unexported functions where it is only being used for cancellation and *never supplemented* with values or timeouts, as in this case.  This is exactly why I brought it up in this situation - who is the intended creator of the context?  If it is only ever the `WsConn` constructor, then it makes perfect sense to have it owned by the WsConn.  If that is not agreeable, then it can be created by the constructor and passed around endlessly, but that would be quite pointless despite what the docs for context state.

Reply:
Another way I would say this is that the guidelines in the context documentation are focused on API design.  If some random external consumer calls Run, providing the context, or even provides context as the first argument of the constructor, *that is where* it makes sense to have it passed as an argument.  All of the internals functions run as goroutines that will need to check ctx.Done are an entirely different matter IMO.

Reply:
I get what the author of the context docs is trying to do, but saying you can't store a `Context` as a field is ludicrous. It makes perfect sense to store a `Context` as a field if it simplifies the API or makes the code easier to read without sacrificing intention. If the `Context` is not going change for the lifetime of the `WsConn`, I would pass it to the constructor and store it as a field.

Of course the fact that I agree with @chappjc is probably the reason that you will find that pattern already used quite a bit. There is no requirement that we all do everything the same, but `Context` fields set in the constructor are preferable in many cases, IMO.

Reply:
I do understand the perspective you are looking at it from and I have no problem with making the change. I can see it would be cleaner to have wsconn's constructor call an unexported `Run` and pass the context to it there instead of having the wallet client call `Run` directly and pass the context to it.

Reply:
In any case, I really was just after a dialog about the API design, which happens to be directly related (IMO) to when following such guidelines really is important vs. unnecessarily tedious.

This is probably a discussion for @buck54321 to pick up however, since it's you two guys working on the client right now:  What feels like the most natural way to deal with a WsConn? Is it to let it start on construction or is there a reason why we might want to call Run at some point after construction (I don't know).  A related question is, is there utility in being able to shutdown and then restart a connection, or would the connection manager make a completely new one.